### DISCUSSION TITLE
Web Playground with WASM

### DISCUSSION THREAD

**Discussion by waynexia** [apache/datafusion#9834] [Show and tell]
Hi! I've made a WASM binding for DataFusion (https://github.com/waynexia/datafusion-wasm) and a Web Playground based on it ([repo](https://github.com/waynexia/datafusion-playground), [demo](https://waynexia.github.io/datafusion-playground/))

<img width="1919" alt="datafusion-playground-demo" src="https://github.com/apache/arrow-datafusion/assets/15380403/e2f251ac-9a7d-428d-b907-4b6d7de88287">

It now supports basic queries (like `create`, `insert`, `select` and `explain`). Generally speaking, DataFusion works pretty well in WASM!

My definition of this Web Playground is something like [DuckDB WASM](https://duckdb.org/docs/api/wasm/overview.html), or the `datafusion-cli` in browser. But not sure about what the shape of `datafusion-wasm` binding should be. I.e., expose as many APIs as possible like an embedded computing library, or just focus on the core SQL/DataFrame interface. I would like to hear feedback from you.

**Comment by alamb** [comment]
>  But not sure about what the shape of datafusion-wasm binding should be. I.e., expose as many APIs as possible like an embedded computing library, or just focus on the core SQL/DataFrame interface. I would like to hear feedback from you.


My opinion is that the WASM bindings have two primary usecases:
1. "embedded analytics" in a webapp (aka running SQL or dataframe queries to drive a UI without requiring a backend server connection)
2. Computation on edge (e.g. CloudFlare's recently announced [`Workers`](https://developers.cloudflare.com/workers/) APIL https://developers.cloudflare.com/workers/runtime-apis/webassembly/

Thus I would personally suggest focusing first on the SQL API (as it has the lowest surface area) and then maybe the dataframe API

**Reply by waynexia** [reply]
Thanks for your insight 👍 My personal short-term plan for this binding would look like:
- [ ] Expose configuration APIs (`SessionContext`)
- [ ] Define and expose `RecordBatch`
- [ ] Support GCS and Azure Bolb
- [ ] Support more kinds of queries (fix them in the "upstream" datafusion)

I'll open a tracking issue in the binding repo for easier collaboration.

**Reply by waynexia** [reply]
FYI https://github.com/waynexia/datafusion-wasm/issues/2

**Comment by alamb** [comment]
BTW: What would you think about moving your playground to https://github.com/datafusion-contrib (I can set you up a repo there with full admin rights) ? I think that might allow more people to find and potentially contribute to your project if that is something you are interested with

**Reply by Omega359** [reply]
Another use case for WASM is for UDF's for systems where the UDF can't be 'in-process' because of security concerns. UDF's used in that manner would definitely have some overhead involved because of memory copying, etc but I know of other systems where the UDF's are WASM based.

**Reply by alamb** [reply]
That is a good point - I suspect that is why CloudFlare's Workers use WASM

**Reply by Omega359** [reply]
Singlestore as well - https://docs.singlestore.com/cloud/reference/code-engine-powered-by-wasm/create-wasm-udfs/

**Reply by waynexia** [reply]
Just came back from holiday, sorry for the delay.

>What would you think about moving your playground to https://github.com/datafusion-contrib 

I'd like to! There are two repos, one for the Web playground and one for the WASM binding. Maybe we need two positions in datafusion-contrib?

**Reply by alamb** [reply]
awesome @waynexia  what should they be called? `datafusion-wasm-playground` and `datafusion-wasm-bindings` (to mirror the name of python bindings)?

**Reply by waynexia** [reply]
Looks good to me :+1: I'll transfer the code tomorrow

**Reply by alamb** [reply]
New repos created:

https://github.com/datafusion-contrib/datafusion-wasm-playground
https://github.com/datafusion-contrib/datafusion-wasm-bindings

🚀

**Comment by g3blv** [comment]
Hi @waynexia . Super nice project! I've been looking at your project and trying to build something similar. I'm running into some performance issues that I noticed you've handled better in your datafustion playground.

I'm building with Rust on both sides - a wasmtime host that loads a WASM module (also in Rust, using DataFusion). I've got a .wit file for the interface and I'm building with `cargo component build --release`.

The thing is, my WASM file ends up at 46MB, and wasmtime takes about 3 minutes just to get the native code ready to run. I was poking around your playground and noticed your WASM file is quite a bit smaller (32MB) and runs way faster.

Any chance you could share what you did to get those better numbers?

**Reply by alamb** [reply]
I have noticed that the datafusion-cli binary is also substantially larger these days. Maybe they are related 🤔  I have been thinking of filing a ticket to investigate / look into this code size

**Reply by alamb** [reply]
I filed https://github.com/apache/datafusion/issues/13813 to collect things to improve, including binary size.

**Reply by waynexia** [reply]
For the building part, I prefer using [`wasm-pack`](https://github.com/rustwasm/wasm-pack) toolchain which will contain an optimization step to reduce size. The artifact after it is ~30MB. Plus, when published to npm it will be compressed to reduce network traffic, this can further reduce the size to <10MB and make it generic and usable in the browser.

I'm not using wasmtime, there might be some differences on different target hosts (mine is focused on `wasm32-unknown-unknown`). But the principle should be similar:`cargo build --release` won't produce the optimal size, we need an extra step to strip anything we don't need

**Comment by alamb** [comment]
I am trying to consolidating various WASM related improvements here: 
- https://github.com/apache/datafusion/issues/13815

Let's move additional discussion there

**Comment by alamb** [comment]
@findepi  filed a ticket to discuss making a playground like this part of DataFusion
- https://github.com/apache/datafusion/issues/13818
